fixtures/flight-vite/server/global.js

```diff
@@ -0,0 +1,191 @@
+'use strict';
+
+// This is a server to host CDN distributed resources like module source files and SSR
+
+const path = require('path');
+const url = require('url');
+
+const fs = require('fs').promises;
+const compress = require('compression');
+const chalk = require('chalk');
+const express = require('express');
+const http = require('http');
+const vite = require('vite');
+
+const {renderToPipeableStream} = require('react-dom/server');
+const {createFromNodeStream} = require('react-server-dom-vite/client');
+
+const moduleBasePath = new URL('../src', url.pathToFileURL(__filename)).href;
+
+async function createApp() {
+  const app = express();
+
+  app.use(compress());
+
+  function request(options, body) {
+    return new Promise((resolve, reject) => {
+      const req = http.request(options, res => {
+        resolve(res);
+      });
+      req.on('error', e => {
+        reject(e);
+      });
+      body.pipe(req);
+    });
+  }
+
+  let getClientAsset;
+
+  if (process.env.NODE_ENV === 'development') {
+    const vite = await import('vite');
+    const {default: reactRefresh} = await import('@vitejs/plugin-react');
+    const viteServer = await vite.createServer({
+      appType: 'custom',
+      server: {middlewareMode: true},
+      plugins: [
+        reactRefresh(),
+        {
+          name: 'react-server-dom-vite:react-refresh',
+          handleHotUpdate({file}) {
+            // clear vite module cache so when its imported again, we will
+            // get the new version
+            globalThis.__vite_module_cache__.delete(file);
```

Comment:
@gnoff is working on a general solution to this problem as it’s a problem in the Webpack impl too atm. Next.js works around it.

Reply:
What kind of general solution are we thinking about? Do I need to do something to prepare for that?

Reply:
React's fast refresh currently runs synchronously but it should probably be a transition. In that transition application code needs to be able to run. For instance refreshing the router. We may need a way to provide modules in a transition compatible way.

I don't yet know what the eventual solution will look like just that there will likely be some API to hook into a refresh and maybe provide a new module cache or something like component families but for modules

You don't need to do anything now to prepare per se, but when I get started on it I'll reach out to figure out what needs to change

Reply:
It turns out that I think we probably don't need that approach after https://github.com/facebook/react/pull/26985 since we have another way to avoid the cache.

One possible solution is to use the metadata object as a key (or just stashing the value on the metadata) because that object is unique to an RSC request. It means it'll go through the async pass during new RSC loads even for existing modules but that's only during the preload phase and not rendering.

Reply:
I'm not sure I get why this is needed for Vite though since Vite adds a timestamp for the new version of the module. If that timestamp is not used by the RSC loader, I suspect that leads to other issues so this might indicate another bug.